### DISCUSSION TITLE
How to setup ddev for a company-like way?

### DISCUSSION THREAD

**Discussion by iammati** [ddev/ddev#2686] [General]
I've got the "permission" to re-build our intern infrastructure (we are using at the moment VMWare + Docker) which is very slow + apache webservers and it's a pain in the a** switching PHP versions, starting the dockers just to start developing...

I've used ddev for longer than a month now and I'd like to have it also at my work. Now the question, how should I set it up?
We got at work a VM1 (intern name of our virtual machine) where each employee/developer got his own "docker" (?) who can then create inside of them for each project a new container (or vice-versa, I'm not really that much into docker, probably only the basics and little more). Say we got a new customer for the project xy, I'd then create a new docker called "xy.myname.lan" (could be xy.myname.ddev or something similiar).

Say I'd flush all the data on VM1 or make a new VM2, install ddev for each employee/dev and they handle every project by themselves using ddev config (to kick-start a project) and such. Would that be the right way or is there something "better" / what I probably could improve?

Greets and thanks for reading!

**Comment by rfay** [comment]
Really ddev is intended to be used by each person on their own computer. Most teams work that way.

**Reply by iammati** [reply]
Oh man, thought there might be some more generalized way to manage all and such. Thanks for that information though.

**Comment by rfay** [comment]
I think you'll like it! It's pretty generalized. You check in the configuration of a project with the project, and anybody on any platform can just `ddev start`

**Comment by iammati** [comment]
Would it be possible to make then some sort of master-daemon server communication where master would be a "big" server which allows developers (daemons) connect to that which then decides what to do e.g.

1. Daemon connects to my VM1 (big master server) with the data that i want to create a new project (ddev config for the project e.g. 'tesla')
2. Master server (VM1) would then make a ddev config in a new folder with the 'tesla' name and handles a configuration provided by the daemon at all etc. OR for each dev its own "sub VM" on that one big master VM?

And sorry, the question is not really ddev-related but I'd like to know some opinions about that which would be a good starting point

PS: I do like it, that's why I'm trying to figure some way to manage it using ddev! :)

**Comment by rfay** [comment]
I think you're trying to fit ddev (and really all local development) into an old paradigm which people have moved away from for good reason. 

The reason for doing development locally is so that you can do your own thing with your own resources without affecting anybody else. Then once you test it, you can push the code, and then lots of people will test on a staging environment, a controlled place that doesn't have everybody competing for space on it.

There are a few ddev users on macOS who run a Linux machine for performance reasons, and ssh into the linux machine. But that's just a different variant on local development (using an extra computer for it).

So if you can help me understand the reason you're trying to move *away* from the idea of "local development" I may be able to help more.

**Comment by iammati** [comment]
It's more like I don't want to move away from local development but more likely that I'm forced to.

So I got on my local laptop ddev running normal. Got some projects on it and push my changes to a remote repository for a staging & afterwards if its fine on a live environment. All fine.

Now, at my company, we would like to have this in combination of another server which hosts our VMs (each dev got his own VM).
The purpose of this additional server, which hosts all the VMs of all developers, is that we can have access to the files of another user.
If there's any way to "break" that barrier to let me for example have access to the files and folders (e.g. in the normal file explorer of windows) in case another developer is on vacation or got sick recently.

Of course, developers could always push their changes but might not be efficiently always since some be for example on their way to lets say doctor or in serious cases in a hospital and such etc... that would not be that productive if you know what I mean by that.

So, TL:DR; Let developer-1 have access to the file mount of developer-2: Possible or not? And if not, any possibility to make it still working in literally ANY way? I'm honestly open for all.

**Comment by rfay** [comment]
Are you talking about the code? Code is managed with source control. So one dev pushes their code to github or gitlab or wherever, and people collaborate with each other that way. Highly recommended. If you're not already working with your code in git, that's the very first thing you need to do.

**Comment by iammati** [comment]
I mean, we do that already no worries about that. I mean more to have access to the files of another developer. Say we got our developer-1 who started a new project but didn't made a init commit yet nor pushed the new project to the remote repository but has finished the base for the project. How would developer-2 be able now to get access to his files in order to make file changes, initialize git and push to a remote repository?

**Comment by rfay** [comment]
* Teach them to check in at the end of the day :) 
* Have each developer share their code from their computer (not that good an idea)

It's time to come into the world of source control I think. Mostly people don't ever expect anybody to be fiddling with their code when they're out for the day. Changes should be made by pull requests, and that can be to a feature branch if a team is working together on it. Your life will be way better (and your team's life) if you embrace modern source control techniques.

**Comment by iammati** [comment]
Yes we do that already. Like primary, we work using git + an intern hosted git-server to manage all the time. The problem is that there is still the possibility that e.g. some files/folders are in the .gitignore and then you require e.g. a .env-file which isn't included in the repository. Of course documentation for a project is also important e.g. what a .env-file requires and such using a .env.example or something similiar. But that should be only an example.

I totally get your point. But having access to the filemount or at least one public folder (say /home/dev/projects/-folder) where all projects are located would be fine or just some SSH tunneling-way. I think that's more possible rather than using the file explorer and such, is that correct?

**Comment by rfay** [comment]
If everybody is in the same network, you could certainly have them work on a network share. You could also have them share the directory where they work.

None of this requires you to think about changing how you use ddev, or how ddev projects are set up.

**Comment by iammati** [comment]
Yes that's perfect then. We're all in the same network. When peoples are in home office we use normally VPN to have access to custom intern subdomains (ticket tool, git etc) so I guess that's also totally fine. I'll try it out and will write here back on Monday if I got something new. Thanks a lot for DDEV and for your help here! Really appreciated it

**Comment by iammati** [comment]
@rfay Okay so got the following problem when trying to install ddev on a debian VM:
When executing brew tap drud/ddev && brew install ddev it stucks to the part when installing nss - waited now +20min and also tried to cancel and re-try it. No luck.

Any idea what could cause that?
![Screenshot 2020-12-12 113603](https://user-images.githubusercontent.com/41418763/101981595-6f852d00-3c6e-11eb-8277-5fc3abf288ef.png)

Idea was to install ddev on the external VM rather on local using Docker Desktop.

**Comment by iammati** [comment]
![image](https://user-images.githubusercontent.com/41418763/101981968-de638580-3c70-11eb-95f6-dc6f91c4e5a8.png)
Okay just got that a couple of minutes ago. Should be reported at homebrew-ddev but I've post it before already here.. ups

**Reply by rfay** [reply]
Please show me the output of `lsb_release -a` and `uname -a` please.

**Reply by rfay** [reply]
If you're on arm64... well Linuxbrew doesn't yet support arm64 and gives lots of messages to that effect.

But I guess I'd probably suggest just using the shell install technique, with install_ddev.sh, it's massively faster than homebrew anyway.

**Reply by rfay** [reply]
I don't have any trouble with a Debian 10 (amd64) `brew upgrade ddev`

**Comment by iammati** [comment]
![image](https://user-images.githubusercontent.com/41418763/101986721-eed72880-3c8f-11eb-8d3e-2c234b5790af.png)
![image](https://user-images.githubusercontent.com/41418763/101986725-f4347300-3c8f-11eb-80bb-1228a4ecf091.png)

Results of lsb_release & uname

**Reply by rfay** [reply]
Just so you know... That's an Ubuntu VM, not Debian. Never had any trouble with Ubuntu amd64 and homebrew, no idea what you have going.

**Reply by iammati** [reply]
Ohh okay. I don't know much about sys-admin related things. More into development of applications rather than server things, good to know!

**Comment by iammati** [comment]
Okay, the install_ddev.sh script (from here https://github.com/drud/ddev/blob/master/scripts/install_ddev.sh) works fine.
I'll try giving it a shot now using the Debian VM - I hope it still got fast enough performance ^^

**Comment by iammati** [comment]
@rfay I'm not sure how ddev connects to docker, but running the `docker` command works fine. It displays me the normal help list. Any idea?
![image](https://user-images.githubusercontent.com/41418763/101987341-cea96880-3c93-11eb-8193-3ef7eb320589.png)

**Reply by rfay** [reply]
Please read the installation instructions, https://ddev.readthedocs.io/en/stable/users/docker_installation/#linux-installation-docker-and-docker-compose - on Ubuntu 20.04 it's just `sudo apt-get install docker.io docker-compose`

Please don't forget the post-installation requirements. You have to add your user to the docker group. It looks like perhaps you haven't done that.

**Reply by iammati** [reply]
Getting an error now - guess it's a docker-/VM-sided issue with the network?
![image](https://user-images.githubusercontent.com/41418763/101988456-ff8c9c00-3c99-11eb-9f60-31d89d953eec.png)

Finished also the installation. DDEV is installed now..

**Reply by rfay** [reply]
You're having difficulty communicating with hub.docker.com. Try again. You may have network problems, who knows.

**Comment by iammati** [comment]
Hey @rfay 

so I've had some tests done of the performance using ddev on the Ubuntu-VM (note that it's a real small .iso-file the VM is running with, like its just the CLI, nothing more) and that things still require a lot of time.

Guess it'd be way better using it, as you mentioned already, ddev on local computers running where the employees use WSL2 to let the dockers run on the VM.

Would it be **possible** to make some sort of rsync-script, which e.g. runs twice a day - once in the morning and once around afternoon/evening automatically (say using a cron-job telling windows to execute a command on the Powershell) - to synchronize, besides of git, all files. Of course that'd require the employee to start his PC, but that would be it since everything else would run automatically, would like to hear your opinion about that. It's something which makes our boss at work worry that somebody is stucked at some work, didn't commit (not even a WIP-commit) and is for any reason not able to work.

Greets

**Comment by rfay** [comment]
There are many tools for syncing filesystems. You may want to try [syncthing](https://syncthing.net/) or use a script with rsync. There are so many tools it's too many to count. Use Dropbox maybe. 

But none of this has anything to do with ddev :)

**Reply by iammati** [reply]
Okay true, your point ^^

Will try them out, guess this discussion can be closed (if it can be at all).

Thanks!
